### PR TITLE
MM-23367 Add docs about test metadata on Cypress spec files

### PR DESCRIPTION
**Pull request by saturninoabril**
#### Summary
Add docs about test metadata on Cypress spec files
- Related PR - https://github.com/mattermost/mattermost-webapp/pull/5107

#### Ticket Link
JIRA ticket: https://mattermost.atlassian.net/browse/MM-23367

#### Screenshot
Screen Shot 2020-03-30 at 4 17 55 PM

### CODE REVIEW

File: site/content/contribute/webapp/end-to-end-tests.md [hunk 1]
```diff
@@ -81,7 +81,7 @@ Inside of the `integration` directory, there are additional directories that bre
 
 1. The [recommended practice](https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements) of Cypress is to use `data-*` attribute to provide context to selectors, but we prefer to use element ID instead.
    - If element ID is not present in the webapp, you may add it in `camelCase` form with human readable name (e.g. `<div id='sidebarTitle'>`). Watch out for potential breaking changes in the snapshot of the unit testing.  Run `make test` to see if all are passing, and run `npm run updatesnapshot` or `npm run test -- -u` if necessary to update snapshot testing.
-2. Add commands or shortcuts to `/e2e/cypress/support/commands.js` (e.g. `toAccountSettingsModal`) that makes it easier to access a page, section, modal and etc. by simply using it as `cy.toAccountSettingsModal('user-1')`.
+2. Add commands or shortcuts to `/e2e/cypress/support/commands.js` (e.g. `toAccountSettingsModal`) that makes it easier to access a page, section, modal and etc. by simply using it as `cy.toAccountSettingsModal()`.
```
Comment (saturninoabril) [inline site/content/contribute/webapp/end-to-end-tests.md]: Not related but fix to update E2E docs

File: site/content/contribute/webapp/end-to-end-tests.md [hunk 2]
```diff
@@ -107,17 +107,34 @@ Inside of the `integration` directory, there are additional directories that bre
...
        before(() => {
-           // Use "cy.hasLicenseForFeature" if the test requires a certain licensed feature
-           cy.hasLicenseForFeature('GuestAccounts');
+           // Use "cy.requireLicenseForFeature" if the test requires a certain licensed feature
+           cy.requireLicenseForFeature('GuestAccounts');
 
-           // Otherwise use "cy.hasLicense" to check if server has license in general
-           cy.hasLicense();
+           // Otherwise use "cy.requireLicense" to check if server has license in general
+           cy.requireLicense();
```
Comment (saturninoabril) [inline site/content/contribute/webapp/end-to-end-tests.md]: Not related but fix to update E2E docs

File: site/content/contribute/webapp/end-to-end-tests.md [hunk 3]
```diff
@@ -107,17 +107,34 @@ Inside of the `integration` directory, there are additional directories that bre
...
        }
     }
     ```
 
 6. Refer to [this pull request](https://github.com/mattermost/mattermost-webapp/pull/3020/files) as a guide on how to write and submit an end-to-end testing PR.
 
+### Adding test metadata on spec files
+Test metadata is used to identify each spec file ahead of time before it is forwarded for Cypress run. Currently, supported test metadata are the following:
+1. "Stage" - an environment for testing, e.g. `@prod`, `@smoke`, `@pull_request`. "Stage" metadata is owned and controlled by QA team who carefully analyze stability of test and promote/demote into a certain stage. This is not required when submitting a spec file.
```
Comment (lindalumitchell) [inline site/content/contribute/webapp/end-to-end-tests.md]:
```suggestion
1. "Stage" - indicates environment for testing, e.g. `@prod`, `@smoke`, `@pull_request`. "Stage" metadata is owned and controlled by QA team who carefully analyze stability of test and promote/demote into a certain stage. This is not required when submitting a spec file.
```

File: site/content/contribute/webapp/end-to-end-tests.md [hunk 4]
```diff
@@ -107,17 +107,34 @@ Inside of the `integration` directory, there are additional directories that bre
...
+2. "Group" - means of grouping tests which are primarily based from tabs of Release testing spreadsheet, e.g. `@account_setting` for Account Setting, `@messaging` for Messaging, etc., or could have a different name depending on how the team decide to group a set of tests. This is required when submitting a spec file.
```
Comment (lindalumitchell) [inline site/content/contribute/webapp/end-to-end-tests.md]:
```suggestion
2. "Group" - indicates test group or category, which is primarily based on functional areas and existing release testing groups, e.g. `@account_settings` for Account Settings, `@messaging` for Messaging, etc. This is required when submitting a spec file.
```

File: site/content/contribute/webapp/end-to-end-tests.md (hunk 3 above)
Comment (saturninoabril) [reply site/content/contribute/webapp/end-to-end-tests.md]: Thanks, updated.

File: site/content/contribute/webapp/end-to-end-tests.md (hunk 4 above)
Comment (saturninoabril) [reply site/content/contribute/webapp/end-to-end-tests.md]: Thanks, updated.

**Comment by saturninoabril** [comment]
Thanks @lindalumitchell, updated.

File: site/content/contribute/webapp/end-to-end-tests.md (hunk 1 above)
Comment (justinegeffen) [inline site/content/contribute/webapp/end-to-end-tests.md:84]:
```suggestion
2. Add commands or shortcuts to `/e2e/cypress/support/commands.js` (e.g. `toAccountSettingsModal`) that make it easier to access a page, section, modal and `/etc` by using it as `cy.toAccountSettingsModal()`.
```
Not sure whether this is referring to /etc or "etc" as in "and so on"?

Comment (saturninoabril) [reply site/content/contribute/webapp/end-to-end-tests.md:84]: It's the latter. Should I change it to "and so on"?

Comment (saturninoabril) [reply site/content/contribute/webapp/end-to-end-tests.md]: Fix per comment, thanks @justinegeffen

**Comment by saturninoabril** [comment]
@hanzei Not sure who to ask about the failing check on `ci`.  This PR is ready to merge -- all comments fulfilled and approved.  Can I merge this even with the failing check on `ci`?

**Comment by hanzei** [comment]
@saturninoabril Thanks for checking. Yes, it's good to go. :+1: 

@metanerd I've noticed that `deploy` fails some time for seemingly no reason. Do you have any additional insights?

**Comment by metanerd** [comment]
@hanzei Do you have any more details or a page you can share at which I can get a deeper look?

**Pull request merged**